oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/Traverser.java

```diff
@@ -0,0 +1,173 @@
+     * <p>
+     * In pre-order traversal, the current node is visited first, followed by its children
+     * from left to right. This method creates an iterator that produces tree nodes in this order.
+     *
+     * @param <T> the type of value in the tree nodes
+     * @param root the root node of the tree, may be null
+     * @param childExtractor function to extract children from a node, must not be null
+     * @return an iterator that traverses the tree in pre-order
+     * @throws NullPointerException if childExtractor or any child is null
+     */
+    @NotNull
+    public static <T> FluentIterable<T> preOrderTraversal(final T root, final @NotNull Function<T, Iterable<? extends T>> childExtractor) {
+
+        Objects.requireNonNull(childExtractor, "Children extractor function must not be null");
+
+        if (root == null) {
+            return FluentIterable.empty();
+        }
+
+        return FluentIterable.of(new Iterable<>() {
+            @Override
+            public @NotNull Iterator<T> iterator() {
+                return UnmodifiableIterator.unmodifiableIterator(new PreOrderIterator<>(root, childExtractor));
+            }
+        });
+    }
+
+    private static final class PreOrderIterator<T> implements Iterator<T> {
+
+        private final Deque<Iterator<? extends T>> stack;
+        private final Function<T, Iterable<? extends T>> childExtractor;
+
+        public PreOrderIterator(final T root, final Function<T, Iterable<? extends T>> childExtractor) {
+            this.childExtractor = childExtractor;
+            this.stack = new ArrayDeque<>();
+            // add first element during initialization
+            stack.addLast(Collections.singletonList(root).iterator());
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Remove any empty iterators from the top of the stack
+            while (!stack.isEmpty() && !stack.peek().hasNext()) {
+                stack.pop();
+            }
+            return !stack.isEmpty();
+        }
+
+        @Override
+        public T next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException("No more nodes in the tree");
+            }
+
+            // Get next element from the current iterator
+            T current = stack.peek().next();
+
+            // Push the iterator for children onto the stack
+            // Children added later will be processed first in pre-order traversal
+            Iterator<? extends T> childIter = childExtractor.apply(current).iterator();
+            if (childIter.hasNext()) {
+                stack.push(childIter);
+            }
+            return current;
+        }
+    }
+
+    /**
+     * Returns an iterator that traverses a tree structure in breadth-first order.
+     * Null nodes are strictly forbidden.
+     * <p>
+     * In breadth-first traversal, all nodes at a given level are visited before any nodes
```

Comment:
I understood this as if for

~~~
/a/b/c and /a/d/e
~~~

it would traverse in this order:

~~~
/a
/a/b
/a/d
/a/b/c
/a/d/e
~~~

which is not the case, right?